sites/cheerpj/src/content/docs/13-tutorials/05-interoperability-tutorial.mdx

```diff
@@ -0,0 +1,159 @@
+---
+title: Java and JavaScript Interoperability
+description: Learn how to enable Java and JavaScript communication using CheerpJ.
+---
+
+With CheerpJ, a Java application can integrate with JavaScript in a browser environment, enabling seamless interaction between Java and JavaScript code.
+
+In this tutorial, we’ll explore creating a two-way communication channel between Java and JavaScript using CheerpJ. The idea is to allow a Java class to access JavaScript functions and vice versa, to enable sending data between the two environments.
+
+## Prerequisites
+
+{/* - [Download the template project](/docs/cheerpj3/tutorials/CheerpJInteroperabilityTutorial.zip) and unzip it. */}
+
+- [Download the application files](/docs/cheerpj3/tutorials/CheerpJInteroperabilityTutorial.zip) and extract them.
+- [Node.js](https://nodejs.org/en/) (>= 18)
+
+## Project Structure
+
+The project consists of the following files and directories:
+
+```plaintext
+CheerpJInteroperbailityTutorial/
+├── com/example
+│   └── Example.java
+│   └── manifest.txt
+├── index.html
+└── example.jar
+└── Makefile
+```
+
+## Setting Up the Project
+
+We start by creating a simple Java class with a native method that will interact with JavaScript. This class will receive input from JavaScript, process it, and send a response back to JavaScript.
+
+```java title="Example.java"
+package com.example;
+
+public class Example {
+    public static native void sendToHTML(String s);
+
+    public String processInput(String input) {
+        sendToHTML(input);
+        return "Java received: " + input;
+    }
+}
+```
+
+In the `Example` class, we define a native method `sendToHTML` that will be implemented in JavaScript. The `processInput` method receives input from JavaScript, sends it to the `sendToHTML` method, and returns a response.
+
+## Implementing the JavaScript Side
+
+Next, we create a JavaScript function that receives input from Java, displays it in the HTML document, and logs the input to the console.
+
+```js title="index.html"
+async function Java_com_example_Example_sendToHTML(lib, str) {
+	document.getElementById("javaOutput").innerText =
+		"JavaScript received: " + str;
+	console.log("Received input from Java: " + str);
+}
+```
+
+Here, the `Java_com_example_Example_sendToHTML` function in JavaScript acts as a native method, allowing a Java class to communicate with JavaScript.
+
+## Handling Multiple JavaScript Instances
+
+To establish persistent communication between Java and JavaScript, we’ll set up a long-running Java thread that enables JavaScript to access Java methods continuously. For example, using `cheerpjRunLibrary` to execute a Java class with native methods creates a separate process from `cheerpjRunJar`, which can prevent resources from being shared between instances.
```

Comment:
The title is not completely correct, we are not handling multiple JavaScript instances. This is correct `To establish persistent communication between Java and JavaScript, we’ll set up a long-running Java thread that enables JavaScript to access Java methods continuously.` But I would remove the sentence after to avoid confusion about `cheerpjRunLibrary`. This was important information for you, but is not really needed in the tutorial. We can also move this to the part where we explain all the Java source code.